episodes/extending never/code.ts

```diff
@@ -0,0 +1,262 @@
+//    e.g. `T extends never` is false for all `T`,
+//          but `never extends never` is true
+
+// 3. nothing is assignable to `never`
+//    e.g. `let param: never; param = "a"`
+//          will produce an error
+
+
+/**
+ * always returns `number`,
+ * except when `never` is passed into `T`
+ */
+function assert1<T>(
+  expectTrue: T extends never
+              ? number
+              : number
+) {}
+
+/**
+ * Passing in `never` as the generic argument results in 
+ * 1. the type of `expectTrue` resolving to `never`, and
+ * 2. an error being triggered for all values of the
+ *    function argument `expectTrue`, due to all types
+ *    (other than `never`) lacking assignability to the
+ *    `never` type.
+ */
+assert1<never>(0) // Error: Argument of type 'number' is not assignable to parameter of type 'never'.(2345)
+//^?
+
+/**
+ * always returns `number`,
+ * even if `never` is passed into `T`
+ */
+function assert2<T>(
+  expectTrue: [T] extends [never]
+              ? number
+              : number
+) {}
+/**
+ * Passing in `never` as the generic argument results in 
+ * 1. the type of `expectTrue` resolving to `number`, and
+ * 2. no errors for all `number` type value function arguments
+ */
+assert2<never>(0)
+//^?
+
+
+
+
+
+
+//////////////////////////////////////////
+// ## Case A: Non-Generic Type
+//            `never extends never`
+
+type NeverExtendsNever =
+//  ^?
+  never extends never 
+  ? true
+  : false
+
+type NEN =
+//  ^?
+  Extract<string | null, undefined> extends never 
+  ? true 
+  : false
+
+type Equivalencies_NEN = [
+  (Extract<string | null, undefined>) extends never
+    ? true
+    : false,
+
+  (never) extends never
+    ? true
+    : false,
+
+  true
+]
+
+
+
+
+
+
+//////////////////////////////////////////
+// ## Case B: Distributive Generic Parameter
+//            `T extends never`
+
+/** The `true` branch is **unreachable** */
+type NeverGeneric<T> = T extends never ? true : false   
+type NG = NeverGeneric<never>
+//   ^?
+
+// For comparison:
+//   a conditional that's distributed over a union
+//   of one, single element:
+type Equivalencies_DC2 = [
+  DistributiveConditional<string>,
+
+  DistributiveConditional</* nothing */ | string>,
+
+  | DistributiveConditional<string>,
+
+  | (string extends string ? string[] : never),
+
+  | string[],
+
+  string[],
+]
+
+type NeverGenericEquivalences = [
+  NeverGeneric<never>,
+  
+  /**
+   * `never` is the empty union of ZERO elements
+   * 
+   * NeverGeneric<(|)>,
+   */
+
+  /**
+   * This distribution over the empty union
+   * never happens (there's nothing to distribute over).
+   * 
+   * | NeverGeneric<(|)>
+   * | NeverGeneric<(|)>,
+   */
+
+  /**
+   * This conditional is never evaluated, and
+   * the `true`/`false` branches are both unreachable.
+   * 
+   * ((|) extends never ? true : false),
+   */
+
+  /**
+   * NeverGeneric<never> resolves to an empty union
+   * 
+   * (|),
+   */
+
+  never,
+]
+
```

Comment:
I'd like to make some edits here so there's more of a direct parallel to the single-element union case above (I'm also using parentheses to signal hypotheticals). It's tricky because none of this is valid typescript syntax, but I think this version might be a bit less confusing.

```ts
/** This distribution over the empty union never happens (there's nothing to distribute over).
     * 
     *    (| NeverGeneric<(|)>), 
     */

    /** This conditional is never evaluated, and the `true`/`false` branches are both unreachable.
     * 
     *    (| ((|) extends never ? true : false)),
     */

    /** NeverGeneric<never> resolves to an empty union.
     * 
     *    (|),
     */
  
    /** ...which is `never` by definition. */
    never,
  ]>
  // Note: `(|)` is not a valid TypeScript expression. 

```

Reply:
sure, what've you got in mind?  don't worry too much about the textual representation since when we are recording the video we'll highlight sections of things with our cursors which will make it make a lot more sense.  all that matters is that it's clear in the video at the end